File: problem/Smallest_Subarrays_OR/smallestSubarrays.java
class Solution {
    private boolean isSet(int num, int pos) {
        return (num & (1 << pos)) != 0;
    }

    public int[] smallestSubarrays(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        int[] suf = new int[n];

        suf[n - 1] = nums[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            suf[i] = suf[i + 1] | nums[i];
        }

        int l = 0, r = 0;
        int cur = nums[0];

        int[] freq = new int[32];

        for (int j = 0; j < 32; j++) {
            if (isSet(nums[0], j)) {
                freq[j]++;
            }
        }

        while (l < n) {
            while (r + 1 < n && cur != suf[l]) {
                r++;
                cur |= nums[r];
                for (int j = 0; j < 32; j++) {
                    if (isSet(nums[r], j)) {
                        freq[j]++;
                    }
                }
            }

            res[l] = Math.max(1, r - l + 1);

            for (int j = 0; j < 32; j++) {
                if (isSet(nums[l], j)) {
                    freq[j]--;
                    if (freq[j] == 0) {
                        cur = cur & (~(1 << j));
                    }
                }
            }

            l++;
        }

        return res;
    }
}

File: problem/Smallest_Subarrays_OR/smallestSubarrays.py
class Solution:
    def is_set(self, num, pos):
        return (num & (1 << pos)) != 0

    def smallestSubarrays(self, nums):
        n = len(nums)
        res = [0] * n
        suf = [0] * n

        suf[n - 1] = nums[n - 1]
        for i in range(n - 2, -1, -1):
            suf[i] = suf[i + 1] | nums[i]

        l = r = 0
        cur = nums[0]
        freq = [0] * 32

        for j in range(32):
            if self.is_set(nums[0], j):
                freq[j] += 1

        while l < n:
            while r + 1 < n and cur != suf[l]:
                r += 1
                cur |= nums[r]
                for j in range(32):
                    if self.is_set(nums[r], j):
                        freq[j] += 1

            res[l] = max(1, r - l + 1)

            for j in range(32):
                if self.is_set(nums[l], j):
                    freq[j] -= 1
                    if freq[j] == 0:
                        cur &= ~(1 << j)

            l += 1

        return res